src/main/java/org/springframework/data/elasticsearch/core/SearchHitMappings.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.springframework.data.elasticsearch.core.document.SearchDocument;
+import org.springframework.data.elasticsearch.core.document.SearchDocumentResponse;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.data.mapping.context.MappingContext;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * @author Rizwan Idrees
+ * @author Mohsin Husen
+ * @author Christoph Strobl
+ * @author Peter-Josef Meisch
+ * @author Mark Paluch
+ * @author Roman Puchkovskiy
+ * @since 4.0
+ */
+class SearchHitMappings {
```

Comment:
To be honest, I usually prefer to reduce the number of method parameters. To have one field more is not a big deal, it's somewhere in the 'data' section of the class and almost never gets in your way when you read the code in methods. But each extra parameter clutters the *active* code itself.

This nested structure is not too elegant, but it is hidden/encapsulated in the class where it is defined, and does not irritate the callers. But as a result, the `mappingContext` is encapsulated and does not need to be passed.

Please let me know if this is convincing or not. If not, I will surely follow your advice :)

Reply:
I'd rather have a second method parameter than the nested class construct. But you are right, here this would need to be intorduced in all the methods, becuase the calling ones need to have it as well.
Haveing a closer look, But we don't need neither the additional parameter nor the outer class. When removing the outer `SearchHitMappings` and keeping the inner this way:
```
class SearchHitMapping<T> {

	private final MappingContext<? extends ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty> mappingContext;
	private final Class<T> type;

	static <T> SearchHitMapping<T> mappingFor(Class<T> entityClass,
			MappingContext<? extends ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty> mappingContext) {
		return new SearchHitMapping<>(entityClass, mappingContext);
	}

	private SearchHitMapping(Class<T> type,
			MappingContext<? extends ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty> mappingContext) {
		Assert.notNull(type, "type must not be null");
		Assert.notNull(mappingContext, "mappingContext must not be null");

		this.type = type;
		this.mappingContext = mappingContext;
	}

        // all the other methods
}
```
we have one class and we don't need the additional fields in the template classes; there we then have:
```
public SearchHits<T> doWith(SearchDocumentResponse response) {
	List<T> entities = response.getSearchDocuments().stream()
			.map(delegate::doWith)
			.collect(Collectors.toList());
	return SearchHitMapping.mappingFor(type, elasticsearchConverter.getMappingContext()).mapHits(response, entities);
}
```

Reply:
Yes, I like this too, I've applied changes according to your suggestion. Thanks.